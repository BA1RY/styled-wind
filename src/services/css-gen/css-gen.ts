import last from 'lodash/last';

export class CSSGen {
  // classes which need no regex matching
  private staticClasses = {
    clearfix: `content: "";
               display: table;
               clear: both;`,
    // font smoothing
    antialiased: `-webkit-font-smoothing: antialiased;
                  -moz-osx-font-smoothing: grayscale;`,
    'subpixel-antialiased': `-webkit-font-smoothing: auto;
                             -moz-osx-font-smoothing: auto;`,
    'break-normal': `overflow-wrap: normal;
                      word-break: normal;`,
    'break-words': `overflow-wrap: break-word;`,
    'break-all': `word-break: break-all;`,
    truncate: `overflow: hidden;
                 text-overflow: ellipsis;
                 white-space: nowrap;`,
    'box-border': 'box-sizing: border-box',
    'box-content': 'box-sizing: content-box'
  };

  private propertyValueClasses = [/float-(.*)/, /overflow-(.*)/];
  private dynamicPropertyClasses = [/m(.)?-(.*)/, /p(.)?-(.*)/];
  private config: any;

  constructor(config: any) {
    this.config = config;
  }

  // classes like clearfix, word-wrap etc.
  private hydrateCustomClasses(classes: string[]) {
    const hydratedCSS: string[] = classes.map(
      (className) => this.staticClasses[className] || className
    );

    return hydratedCSS;
  }

  // float-left => flat: left;
  // overflow-x-hidden => overflow-x: hidden;
  private hydratePropertyValueClasess(classes: string[]) {
    const hydratedCSS = classes.map((className) => {
      const propertyValueClass = this.propertyValueClasses.find(
        (propertyValueClass) => propertyValueClass.test(className)
      );

      if (propertyValueClass) {
        // overflow-x-hidden
        // property = overflow, valueWithProperty => x-hidden;
        const [property, ...valueWithProperty] = className.split('-');
        // allProprty = overflow-x
        const allProperties = [
          property,
          ...valueWithProperty.slice(0, valueWithProperty.length - 1)
        ].join('-');

        return `${allProperties}: ${last(valueWithProperty)};`;
      } else {
        return className;
      }
    });

    return hydratedCSS;
  }

  // mt-10, m-10, p-10, pb-10
  private hydrateDynamicPropertyClasses(classes: string[]) {
    const hydratedCSS = classes.map((className) => {
      const dyanmicPopertyClass = this.dynamicPropertyClasses.find(
        (dynamicPropertyClass) => dynamicPropertyClass.test(className)
      );

      if (dyanmicPopertyClass) {
        // switch based on the first letter whether it is m, p etc
        switch (className[0]) {
          case 'm': {
            const [, direction, value] = className.match(dyanmicPopertyClass);

            // TODO: not sure how to use this.config.theme.spacing
            const themeValue = this.config.theme.spacing[value];
            const directionString = this.getDirectionString(direction);

            return `margin${directionString}: ${themeValue};`;
          }
          case 'p': {
            const [, direction, value] = className.match(dyanmicPopertyClass);

            // TODO: not sure how to use this.config.theme.spacing
            const themeValue = this.config.theme.spacing[value];
            const directionString = this.getDirectionString(direction);

            return `padding${directionString}: ${themeValue};`;
          }
        }
      }
    });

    return hydratedCSS;
  }

  private getDirectionString(direction: string) {
    switch (direction) {
      case 't':
        return '-top';
      case 'b':
        return '-bottom';
      case 'r':
        return '-right';
      case 'l':
        return '-left';
    }

    return '';
  }

  genCSS(classes: string[]) {
    const withCustomClasses = this.hydrateCustomClasses(classes);
    const withPropertyValueClasses = this.hydratePropertyValueClasess(
      withCustomClasses
    );
    const withDyanmicValueClasses = this.hydrateDynamicPropertyClasses(
      withPropertyValueClasses
    );

    return withDyanmicValueClasses.join('');
  }
}
