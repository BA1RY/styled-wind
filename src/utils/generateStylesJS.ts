import resolveConfigObjects from './resolveConfig';
import { defaultConfig } from './defaultConfig';

export default function resolveConfig(...configs: any) {
  return resolveConfigObjects([...configs, defaultConfig]);
}
