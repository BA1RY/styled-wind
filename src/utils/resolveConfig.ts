import some from 'lodash-es/some';
import mergeWith from 'lodash-es/mergeWith';
import isFunction from 'lodash-es/isFunction';
import isUndefined from 'lodash-es/isUndefined';
import defaults from 'lodash-es/defaults';
import map from 'lodash-es/map';
import get from 'lodash-es/get';
import toPath from 'lodash-es/toPath';
import negateValue from './negateValue';

const configUtils = {
  negative(scale: any) {
    return Object.keys(scale)
      .filter((key) => scale[key] !== '0')
      .reduce(
        (negativeScale, key) => ({
          ...negativeScale,
          [`-${key}`]: negateValue(scale[key])
        }),
        {}
      );
  },
  breakpoints(screens: any) {
    return Object.keys(screens)
      .filter((key) => typeof screens[key] === 'string')
      .reduce(
        (breakpoints, key) => ({
          ...breakpoints,
          [`screen-${key}`]: screens[key]
        }),
        {}
      );
  }
};

function value(valueToResolve: any, ...args: any) {
  return isFunction(valueToResolve) ? valueToResolve(...args) : valueToResolve;
}

function mergeThemes(themes: any) {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const theme = (({ extend, ...t }: any) => t)(
    themes.reduce((merged: any, t: any) => {
      return defaults(merged, t);
    }, {})
  );

  return {
    ...theme,

    // In order to resolve n config objects, we combine all of their `extend` properties
    // into arrays instead of objects so they aren't overridden.
    extend: themes.reduce((merged: any, { extend }: any) => {
      return mergeWith(merged, extend, (mergedValue: any, extendValue: any) => {
        if (isUndefined(mergedValue)) {
          return [extendValue];
        }

        if (Array.isArray(mergedValue)) {
          return [extendValue, ...mergedValue];
        }

        return [extendValue, mergedValue];
      });
    }, {})
  };
}

function mergeExtensions({ extend, ...theme }: any) {
  return mergeWith(theme, extend, (themeValue: any, extensions: any) => {
    // The `extend` property is an array, so we need to check if it contains any functions
    if (!isFunction(themeValue) && !some(extensions, isFunction)) {
      return {
        ...themeValue,
        ...Object.assign({}, ...extensions)
      };
    }

    return (resolveThemePath: any, utils: any) => ({
      ...value(themeValue, resolveThemePath, utils),
      ...Object.assign(
        {},
        ...extensions.map((e: any) => value(e, resolveThemePath, utils))
      )
    });
  });
}

function resolveFunctionKeys(object: any) {
  const resolveThemePath = (key: any, defaultValue: any) => {
    const path = toPath(key);

    let index = 0;
    let val = object;

    while (val !== undefined && val !== null && index < path.length) {
      val = val[path[index++]];
      val = isFunction(val) ? val(resolveThemePath, configUtils) : val;
    }

    return val === undefined ? defaultValue : val;
  };

  return Object.keys(object).reduce((resolved, key) => {
    return {
      ...resolved,
      [key]: isFunction(object[key])
        ? object[key](resolveThemePath, configUtils)
        : object[key]
    };
  }, {});
}

function extractPluginConfigs(configs = []) {
  let allConfigs: any = [];

  configs.forEach((config) => {
    allConfigs = [...allConfigs, config];

    const plugins = get(config, 'plugins', []);

    if (plugins.length === 0) {
      return;
    }

    plugins.forEach((plugin: any) => {
      if (plugin.__isOptionsFunction) {
        plugin = plugin();
      }
      allConfigs = [
        ...allConfigs,
        ...extractPluginConfigs([get(plugin, 'config', {})] as any)
      ];
    });
  });

  return allConfigs;
}

export default function resolveConfig(configs: any) {
  const allConfigs = extractPluginConfigs(configs);

  return defaults(
    {
      theme: resolveFunctionKeys(
        mergeExtensions(
          mergeThemes(map(allConfigs, (t) => get(t, 'theme', {})))
        )
      ),
      variants: ((firstVariants) => {
        return Array.isArray(firstVariants)
          ? firstVariants
          : defaults({}, ...map(allConfigs, 'variants'));
      })(defaults({}, ...map(allConfigs)).variants)
    },
    ...allConfigs
  );
}
