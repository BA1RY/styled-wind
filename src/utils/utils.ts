import { name } from '../../package.json';

export function warn(message: string) {
  if (
    process.env.NODE_ENV === 'development' ||
    process.env.NODE_ENV === 'test'
  ) {
    console.warn(`${name}: warning: ${message}`);
  }
}
