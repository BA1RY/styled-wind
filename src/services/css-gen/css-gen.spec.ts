import { CSSGen } from './css-gen';
import { defaultConfig } from '../../utils/defaultConfig';

describe('CSSGen', () => {
  let cssGen: CSSGen;

  beforeEach(() => {
    cssGen = new CSSGen(defaultConfig);
  });

  it('should generate css for classes like `clearfix`', () => {
    const css = cssGen.genCSS(['clearfix', 'antialiased']);

    expect(css).toMatchSnapshot();
  });

  it('should generate css for classes like `float-right`', () => {
    const css = cssGen.genCSS(['float-right', 'overflow-x-auto']);

    expect(css).toMatchSnapshot();
  });

  it('should generate css for classes like `mt-10`', () => {
    const css = cssGen.genCSS(['mt-5', 'm-0', 'pt-1', 'p-0']);

    expect(css).toMatchSnapshot();
  });
});
