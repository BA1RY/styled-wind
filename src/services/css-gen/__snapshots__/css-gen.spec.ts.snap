// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSSGen should generate css for classes like \`clearfix\` 1`] = `""`;

exports[`CSSGen should generate css for classes like \`float-right\` 1`] = `""`;

exports[`CSSGen should generate css for classes like \`mt-10\` 1`] = `"margin-top: 1.25rem;margin: 0;padding-top: 0.25rem;padding: 0;"`;
